File: ios/Classes/FlutterRealmPlugin.m
#import <Realm/Realm.h>
#import <Realm/RLMRealm_Dynamic.h>
#import "FlutterRealmPlugin.h"
#import "MethodSubHandler.h"
#import "SyncUserMethodSubHandler.h"
#import "FlutterRealm.h"
#import "FlutterError+NSError.h"

static NSString *const CHANNEL_NAME = @"plugins.it_nomads.com/flutter_realm";


@interface FlutterRealmPlugin ()
@property FlutterMethodChannel *channel;
@property NSMutableDictionary<NSString *, FlutterRealm *> *realms;
@property NSArray<MethodSubHandler> *handlers;
@end


@implementation FlutterRealmPlugin

- (instancetype)initWithChannel:(FlutterMethodChannel *)channel {
    self = [super init];
    _channel = channel;
    _realms = [NSMutableDictionary dictionary];
    _handlers = (NSArray<MethodSubHandler> *)@[[[SyncUserMethodSubHandler alloc] init]];
    return self;
}

+ (void)registerWithRegistrar:(nonnull NSObject<FlutterPluginRegistrar> *)registrar {
    
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:CHANNEL_NAME
                                     binaryMessenger:[registrar messenger]];
    FlutterRealmPlugin* instance = [[FlutterRealmPlugin alloc] initWithChannel:channel];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call result:(FlutterResult)result {
    NSDictionary *arguments = [call arguments];
    NSString *method = [call method];

    @try {
        for (id<MethodSubHandler> handler in self.handlers){
            if ([handler handleMethodCall:call result:result]){
                return;
            }
        }
        if ([@"initialize" isEqualToString:method]){
            NSString *realmId = arguments[@"realmId"];
            NSAssert([realmId isKindOfClass:[NSString class]], @"String realmId must be provided. Got: %@", realmId);
            FlutterRealm *flutterRealm = [[FlutterRealm alloc] initWithArguments:arguments channel:self.channel identifier:realmId];
            self.realms[realmId] = flutterRealm;
            
            result(nil);
        }  else if ([@"reset" isEqualToString:method]){
            for (FlutterRealm *realm in self.realms.allValues){
                [realm reset];
            }
            
            [self.realms removeAllObjects];
            
            result(nil);
        } else if ([@"asyncOpenWithConfiguration" isEqualToString:method]){
            [self handleAsyncOpenWithConfiguration:arguments result:result];
        } else if ([@"syncOpenWithConfiguration" isEqualToString:method]){
            [self handleSyncOpenWithConfiguration:arguments result:result];
        }else {
            NSString *realmId = arguments[@"realmId"];
            NSAssert([realmId isKindOfClass:[NSString class]], @"String realmId must be provided. Got: %@", realmId);
            FlutterRealm *realm = self.realms[realmId];

            if (realm != nil){
                [realm handleMethodCall:call result:result];
            }else {
                result([FlutterError errorWithCode:@"-1" message:@"Realm not found" details:[NSString stringWithFormat:@"Method %@:%@", method, arguments]]);
            }
        }
        
    } @catch (NSException *exception) {
        NSLog(@"%@", exception.callStackSymbols);
        result([FlutterError errorWithCode:@"-1" message:exception.reason details:[exception.userInfo description]]);
    }
}

- (void)handleAsyncOpenWithConfiguration:(NSDictionary *)arguments result:(FlutterResult)result {
    NSString *realmId = arguments[@"realmId"];
    NSString *syncServerURLString = arguments[@"syncServerURL"];
    bool fullSynchronization = [arguments[@"fullSynchronization"] boolValue];
    NSURL *syncServerURL = [NSURL URLWithString: syncServerURLString];
    RLMSyncUser *user = [RLMSyncUser currentUser];
    RLMRealmConfiguration *config = [user configurationWithURL:syncServerURL fullSynchronization:fullSynchronization];
    
    [RLMRealm asyncOpenWithConfiguration:config
                           callbackQueue:dispatch_get_main_queue()
                                callback:^(RLMRealm *realm, NSError *error) {
                                    if (realm) {
                                        FlutterRealm *flutterRealm = [[FlutterRealm alloc] initWithRealm:realm channel:self.channel identifier:realmId];
                                        self.realms[realmId] = flutterRealm;
                                        result(nil);
                                    } else {
                                        result([FlutterError fromNSError:error]);
                                    }
                                }];
}

- (void)handleSyncOpenWithConfiguration:(NSDictionary *)arguments result:(FlutterResult)result {
    NSString *realmId = arguments[@"realmId"];
    NSString *syncServerURLString = arguments[@"syncServerURL"];
    bool fullSynchronization = [arguments[@"fullSynchronization"] boolValue];
    NSURL *syncServerURL = [NSURL URLWithString: syncServerURLString];
    RLMSyncUser *user = [RLMSyncUser currentUser];
    RLMRealmConfiguration *config = [user configurationWithURL:syncServerURL fullSynchronization:fullSynchronization];
    
    NSError *error;
    RLMRealm *realm = [RLMRealm realmWithConfiguration:config error:&error];
    if (error != nil) {
        FlutterError *flutterError = [FlutterError fromNSError:error];
        result(flutterError);
        return;
    }
    
    FlutterRealm *flutterRealm = [[FlutterRealm alloc] initWithRealm:realm channel:self.channel identifier:realmId];
    self.realms[realmId] = flutterRealm;
    result(nil);
}

@end

File: android/src/main/java/com/it_nomads/flutter_realm/FlutterRealmPlugin.java
package com.it_nomads.flutter_realm;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import io.realm.Realm;
import io.realm.SyncConfiguration;
import io.realm.SyncUser;

public class FlutterRealmPlugin implements MethodCallHandler {

    private FlutterRealmPlugin(MethodChannel channel) {
        this.channel = channel;
        handlers = new ArrayList<>();
        handlers.add(new SyncUserMethodSubHandler());
    }

    public static void registerWith(Registrar registrar) {
        Realm.init(registrar.context());

        final MethodChannel channel = new MethodChannel(registrar.messenger(), "plugins.it_nomads.com/flutter_realm");

        FlutterRealmPlugin plugin = new FlutterRealmPlugin(channel);
        channel.setMethodCallHandler(plugin);
    }

    private HashMap<String, FlutterRealm> realms = new HashMap<>();
    private final MethodChannel channel;
    private List<MethodSubHandler> handlers;

    @Override
    public void onMethodCall(MethodCall call, Result result) {

        try {
            Map arguments = (Map) call.arguments;

            for (MethodSubHandler handler : handlers) {
                if (handler.onMethodCall(call, result)) {
                    return;
                }
            }

            switch (call.method) {
                case "initialize": {
                    onInitialize(result, arguments);
                    break;
                }
                case "reset":
                    onReset(result);
                    break;
                case "asyncOpenWithConfiguration":
                    onAsyncOpenWithConfiguration(arguments, result);
                    break;
                case "syncOpenWithConfiguration":
                    onSyncOpenWithConfiguration(arguments, result);
                    break;
                default: {
                    String realmId = (String) arguments.get("realmId");
                    FlutterRealm flutterRealm = realms.get(realmId);
                    if (flutterRealm == null) {
                        String message = "Method " + call.method + ":" + arguments.toString();
                        result.error("Realm not found", message, null);
                        return;
                    }

                    flutterRealm.onMethodCall(call, result);
                    break;
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            result.error(e.getMessage(), e.getMessage(), e.getStackTrace().toString());
        }
    }

    private void onInitialize(Result result, Map arguments) {
        String realmId = (String) arguments.get("realmId");
        FlutterRealm flutterRealm = new FlutterRealm(channel, realmId, arguments);
        realms.put(realmId, flutterRealm);
        result.success(null);
    }

    private void onReset(Result result) {
        for (FlutterRealm realm : realms.values()) {
            realm.reset();
        }
        realms.clear();
        result.success(null);
    }

    private void onAsyncOpenWithConfiguration(Map arguments, final Result result) {
        final String realmId = (String) arguments.get("realmId");
        final SyncConfiguration configuration = getSyncConfiguration(arguments);

        Realm.getInstanceAsync(configuration, new Realm.Callback() {
            @Override
            public void onSuccess(Realm realm) {
                FlutterRealm flutterRealm = new FlutterRealm(channel, realmId, realm);
                realms.put(realmId, flutterRealm);
                result.success(null);
            }

            @Override
            public void onError(Throwable exception) {
                result.error(exception.getLocalizedMessage(), exception.getMessage(), exception);
            }

        });

    }

    private void onSyncOpenWithConfiguration(Map arguments, Result result) {
        String realmId = (String) arguments.get("realmId");
        SyncConfiguration configuration = getSyncConfiguration(arguments);

        FlutterRealm flutterRealm = new FlutterRealm(channel, realmId, configuration);
        realms.put(realmId, flutterRealm);
        result.success(null);
    }

    private SyncConfiguration getSyncConfiguration(Map arguments) {
        String syncServerURL = (String) arguments.get("syncServerURL");
        boolean fullSynchronization = (boolean) arguments.get("fullSynchronization");

        assert syncServerURL != null;

        SyncConfiguration.Builder builder = SyncUser.current().createConfiguration(syncServerURL);
        if (fullSynchronization) {
            builder.fullSynchronization();
        }


        return builder.build();
    }


}
